presto-orc/src/main/java/com/facebook/presto/orc/OrcWriter.java

```diff
@@ -483,16 +604,64 @@ public void close()
         Map<String, Slice> userMetadata = this.userMetadata.entrySet().stream()
                 .collect(Collectors.toMap(Entry::getKey, entry -> utf8Slice(entry.getValue())));
 
+        ImmutableList.Builder<ColumnStatistics> unencryptedStatsBuilder = ImmutableList.builder();
+        ImmutableMultimap.Builder<Integer, ColumnStatistics> encryptedStatsBuilder = ImmutableMultimap.builder();
+        for (int i = 0; i < fileStats.size(); i++) {
+            if (dwrfEncryptionInfo.getGroupByNodeId(i).isPresent()) {
+                encryptedStatsBuilder.put(dwrfEncryptionInfo.getGroupByNodeId(i).get(), fileStats.get(i));
+            }
+            else {
+                unencryptedStatsBuilder.add(fileStats.get(i));
+            }
+        }
+        List<ColumnStatistics> unencryptedStats = unencryptedStatsBuilder.build();
+        Multimap<Integer, ColumnStatistics> encryptedStats = encryptedStatsBuilder.build();
+        Optional<DwrfEncryption> dwrfEncryption;
+        if (dwrfWriterEncryption.isPresent()) {
+            ImmutableList.Builder<EncryptionGroup> encryptionGroupBuilder = ImmutableList.builder();
+            List<WriterEncryptionGroup> writerEncryptionGroups = dwrfWriterEncryption.get().getWriterEncryptionGroups();
+            for (int i = 0; i < writerEncryptionGroups.size(); i++) {
+                WriterEncryptionGroup group = writerEncryptionGroups.get(i);
+                List<byte[]> columnStatistics = encryptedStats.get(i)
+                        .stream()
+                        .map(DwrfMetadataWriter::toColumnStatistics)
+                        .map(AbstractMessageLite::toByteArray)
+                        .collect(toImmutableList());
+
+                // TODO using an object output stream does not necessarily produce the same result that proto would
```

Comment:
@zzhao0 how are the stats supposed to be serialized. Presto doesn't use them, and there didn't see a protobuf method for serializing the list.

Reply:
stats are serialized very similar to the encoding and streams, ie. stats for the same encryption groups are put into the same FileStatistics object, and serialize with encryption. Then stored as blob inside the encryption group.

https://www.internalfb.com/intern/diffusion/FBS/browse/master/fbcode/fbjava/bbio/hive-dwrf/proto/orc_proto.proto?commit=f872cc50345082c014885fc81da222524985a30a&lines=191

if you have schema 
struct<a:array<int>, b:struct<a:int,b:int>, c:int>
a and b.b is in group 1
b.a is in group 2
c is not encrypted

you would expect to have:
encryption group 1:
nodes: a, b.b
stats: [a, a.elements, b.b]. (stats is nested)

stats for group2:
nodes:b.a
stats: [b.a]

two things to call out:
1. stats should be ordered the same way as the node. In group1, we have a as the first node, so we should have all stats for the subtree of a listed in front of all others. Alternatively, we can order it like:
nodes: b.b, a
stats: [b.b, a, a.elements]
reader should still be able to load it. But it's actually easier for writer to write a first...

2. file level stats is a list and is implicitly ordered according to schema. So for encrypted nodes, we still need to have one entry in the list except that it should not contain any detailed stats info (min/max/sum), it's ok to have value count and has null

on the reader side, it should merge stats from two places, and present user with a consistent view

Reply:
ok, I missed that there was supposed to be a FileStatistics object added to the dwrfproto definition (it wasn't highlighted as a change to the file format in the design spec).   That was exactly the problem I was having.  I'll add that to the file format and update this.